python/ray/autoscaler/aws/config.py

```diff
@@ -70,6 +75,135 @@ def key_pair(i, region, key_name):
 # Suppress excessive connection dropped logs from boto
 logging.getLogger("botocore").setLevel(logging.WARNING)
 
+_log_info = {}
+
+
+def reload_log_state(override_log_info):
+    _log_info.update(override_log_info)
+
+
+def get_log_state():
+    return _log_info.copy()
+
+
+def _set_config_info(**kwargs):
+    """Record configuration artifacts useful for logging."""
+
+    # todo: this is technically fragile iff we ever use multiple configs
+
+    for k, v in kwargs.items():
+        _log_info[k] = v
+
+
+def _arn_to_name(arn):
+    return arn.split(":")[-1].split("/")[-1]
+
+
+def log_to_cli(config):
+    provider_name = PROVIDER_PRETTY_NAMES.get("aws", "Unknown provider")
+    with cli_logger.group("{} config", provider_name):
+
+        def same_everywhere(key):
+            return config["head_node"][key] == config["worker_nodes"][key]
+
+        tags = {"default": _log_info["head_instance_profile_src"] == "default"}
+        cli_logger.labeled_value(
+            "IAM Profile",
+            "{}",
+            _arn_to_name(config["head_node"]["IamInstanceProfile"]["Arn"]),
+            _tags=tags)
+
+        if same_everywhere("KeyName"):
+            tags = {"default": _log_info["keypair_src"] == "default"}
+            cli_logger.labeled_value(
+                "Key pair (head & workers)",
+                "{}",
+                config["head_node"]["KeyName"],
+                _tags=tags)
+        else:
+            # cannot be default
+            cli_logger.labeled_value("EC2 Key pair (head)", "{}",
+                                     config["head_node"]["KeyName"])
+            cli_logger.labeled_value("EC2 Key pair (workers)", "{}",
+                                     config["worker_nodes"]["KeyName"])
+
+        if same_everywhere("SubnetIds"):
+            tags = {"default": _log_info["head_subnet_src"] == "default"}
+            cli_logger.labeled_value(  # todo: handle plural vs singular?
+                "VPC Subnets (head & workers)",
+                "{}",
+                cli_logger.render_list(config["head_node"]["SubnetIds"]),
+                _tags=tags)
+        else:
+            tags = {"default": _log_info["head_subnet_src"] == "default"}
+            cli_logger.labeled_value(
+                "VPC Subnets (head)",
+                "{}",
+                cli_logger.render_list(config["head_node"]["SubnetIds"]),
+                _tags=tags)
+
+            tags = {"default": _log_info["workers_subnet_src"] == "default"}
+            cli_logger.labeled_value(
+                "VPC Subnets (workers)",
+                "{}",
+                cli_logger.render_list(config["worker_nodes"]["SubnetIds"]),
+                _tags=tags)
```

Comment:
a lot of this handling seems redundant; can you reduce redundancy?

Reply:
fixed